dotnet/src/Connectors/Connectors.Memory.CosmosNoSql/CosmosNoSqlCollection.cs

```diff
@@ -70,29 +71,63 @@ public sealed class CosmosNoSqlCollection<TKey, TRecord> : VectorStoreCollection
     /// <param name="database"><see cref="Database"/> that can be used to manage the collections in Azure CosmosDB NoSQL.</param>
     /// <param name="name">The name of the collection that this <see cref="CosmosNoSqlCollection{TKey, TRecord}"/> will access.</param>
     /// <param name="options">Optional configuration options for this class.</param>
-    public CosmosNoSqlCollection(
-        Database database,
-        string name,
-        CosmosNoSqlCollectionOptions? options = default)
+    public CosmosNoSqlCollection(Database database, string name, CosmosNoSqlCollectionOptions? options = default)
+        : this(new(database.Client, ownsClient: false), _ => database, name, options)
     {
-        // Verify.
         Verify.NotNull(database);
         Verify.NotNullOrWhiteSpace(name);
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="CosmosNoSqlCollection{TKey, TRecord}"/> class.
+    /// </summary>
+    /// <param name="connectionString">Connection string required to connect to Azure CosmosDB NoSQL.</param>
+    /// <param name="databaseName">Database name for Azure CosmosDB NoSQL.</param>
+    /// <param name="collectionName">The name of the collection that this <see cref="CosmosNoSqlCollection{TKey, TRecord}"/> will access.</param>
+    /// <param name="clientOptions">Optional configuration options for <see cref="CosmosClient"/>.</param>
+    /// <param name="collectionOptions">Optional configuration options for <see cref="VectorStoreCollection{TKey, TRecord}"/>.</param>
+    public CosmosNoSqlCollection(string connectionString, string databaseName, string collectionName,
+        CosmosClientOptions? clientOptions = null, CosmosNoSqlCollectionOptions? collectionOptions = null)
+        : this(new ClientWrapper(new CosmosClient(connectionString, clientOptions), ownsClient: true),
+              client => client.GetDatabase(databaseName), collectionName, null)
+    {
+        Verify.NotNullOrWhiteSpace(connectionString);
+        Verify.NotNullOrWhiteSpace(databaseName);
+        Verify.NotNullOrWhiteSpace(collectionName);
+    }
 
+    internal CosmosNoSqlCollection(
+        ClientWrapper clientWrapper,
+        Func<CosmosClient, Database> databaseProvider,
+        string name,
+        CosmosNoSqlCollectionOptions? options)
+    {
         if (typeof(TKey) != typeof(string) && typeof(TKey) != typeof(CosmosNoSqlCompositeKey) && typeof(TKey) != typeof(object))
         {
             throw new NotSupportedException($"Only {nameof(String)} and {nameof(CosmosNoSqlCompositeKey)} keys are supported (and object for dynamic mapping).");
         }
 
-        if (database.Client?.ClientOptions?.UseSystemTextJsonSerializerWithOptions is null)
+        try
+        {
+            this._database = databaseProvider(clientWrapper.Client);
+        }
+        catch (Exception)
+        {
+            // Something went wrong, we dispose the client and don't store a reference.
+            clientWrapper.Dispose();
```

Comment:
Don't we also have to dispose if e.g. one of the ArgumentExceptions below get thrown?

Consider incrementing the ref count only at the very end of the constructor, this way we're sure there can't be any leaks. This is probably also relevant for the other providers?

Reply:
good catch! For the blocks of code that could throw I've implemented following pattern:

```cs
try
{
   // init logic
}
catch (Exception)
{
   client.Dispose();
   throw;
}

this._client = client; // store the reference only if everything worked fine
```

Reply:
I agree that it's ugly, but it has one advantage: we `Dispose` the resource when any form of validation/initialization logic fails. I could move the increment to the end and remove the catch block, but then we would not dispose the resource on error.